refactor(core): Error code for Component Id collision (#49986)
angular/angular: aio/content/errors/NG0912.md

> @name Component ID generation collision

Reviewer: NIT: I think `ID` should always be uppercase. There are several inconsistencies in this doc.

@category runtime
@shortDescription Component ID generation collision detected

@description

> When creating components, Angular generates a unique component ID for each component. Angular uses component information (such as selectors, the number of host bindings, view and component queries, etc) to generate a unique id. However, when two components are similar (often times sharing the same selector), an ID collision can happen. 

Reviewer: ```suggestion
When creating components, Angular generates a unique component ID for each component. This is generated using the Angular component metadata, including but not limited:  selectors, the number of host bindings, class property names, view and content queries. When two components metadata are identical (often times sharing the same selector), an ID generation collision will occur. 
```

Reviewer: @AndrewKushnir thoughts about the above? Cause I’ve seen that you suggested the correct wording.

Component IDs are primarily used for [hydration](guide/hydration) to restore an application state after [server-side rendering](guide/universal). When component id collision happens, this might result in a partial or incorrect state restoration, so it's recommended to resolve them as described below.

** Example of a Component id collision **

```typescript
@Component({
  selector: 'my-component',
  template: 'complex-template',
})
class SomeComponent {}

@Component({
  selector: 'my-component',
  template: 'empty-template',
})
class SomeMockedComponent {}
```
Having these two components defined will trigger a collision and an error message would be logged in a browser console.

@debugging

> The error message includes the class name of the two components whose IDs are colliding. 

Reviewer: ```suggestion
The warning message includes the class name of the two components whose IDs are colliding. 
```

The problem can be resolved using one of the solutions below: 

1. Change the selector of one of the two components. 

```typescript
@Component({

>   selector: 'my-mocked-component',

Reviewer: Not sure if we should add this to the docs, but there is not need to change the selector completely, as in most cases that would be problematic as they might require update certain usages. 

Users can use the `:not` selector to get around this

```diff
- my-component
+ my-component:not(p)
```

Thoughts @AndrewKushnir ?

Reviewer 2: @alan-agius4 looks good 👍 

I was also thinking about an option like this (add an extra selector to convey the meaning):
```
selector: 'my-component,mocked-for-tests',
```
I think we can do either of those in the guide and @JeanMeche can pick the one that looks easier to explain in the docs :)

@name Component ID generation collision
@category runtime
@shortDescription Component ID generation collision detected

@description

> When creating components, Angular tries to generate a unique component ID for each component. When two components are too similar, an ID collision can happen. 

Reviewer 2: ```suggestion
When creating components, Angular generates a unique component ID for each component. Angular uses component information (such as selectors, the number of host bindings, view and component queries, etc) to generate a unique id. However, when two components are similar (often times sharing the same selector), an ID collision can happen. 

Component ids are primarily used for [hydration](guide/hydration) to restore an application state after [server-side rendering](guide/universal). When component id collision happens, this might result in a partial or incorrect state restoration, so it's recommended to resolve them as described below.
```

> ** Reproduction Example **

Reviewer 2: ```suggestion
** Example of a Component id collision **
```

```typescript
@Component({
  selector: 'my-component',
  template: 'complex-template',
})
class SomeComponent {}

@Component({
  selector: 'my-component',
  template: 'empty-template',
})
class SomeMockedComponent {}
```

> Having these two components defined will fire the following warning: 
> 
> <div class="alert is-important">
> 
> Component ID generation collision detected. Components 'SomeComponent' and 'SomeMockedComponent' with selector 'my-component' generated the same component ID. To fix this, you can change the selector of one of those components or add an extra host attribute to force a different ID.
> 
> </div>
> 
> 

Reviewer 2: I'd propose to drop an error message text, so that it doesn't go out of sync with the code. Also, most likely developers would come to this page after seeing the error anyway.
```suggestion
Having these two components defined will trigger a collision and an error message would be logged in a browser console.
```

@debugging

The error message includes the class name of the two components whose IDs are colliding. 

> Several solutions exists to fix this issues: 

Reviewer 2: ```suggestion
The problem can be resolved using one of the solutions below: 
```

1. Change the selector of one of the two components. 

```typescript
@Component({
  selector: 'my-mocked-component',
  template: 'empty-template',
})
class SomeMockedComponent {}
```

> 2. Add a `collision-id` to the host attribute

Reviewer 2: ```suggestion
2. Add an extra host attribute to one of the components.
```

```typescript
@Component({
  selector: 'my-component',
  template: 'complex-template',

>   host: {'collision-id': 'SomeComponent'},

Reviewer 2: The 'collision-id' looks like a special name to resolve conflicts :) Let's may be change it to something more abstract?
```suggestion
  host: {'some-binding': 'some-value'},
```

})
class SomeComponent {}

@Component({
  selector: 'my-component',
  template: 'empty-template',
  host: {'collision-id': 'SomeMockedComponent'},
})
class SomeMockedComponent {}
```

> 3. Change the class signature by adding a method to one of the components. 

Reviewer 2: ```suggestion
3. Change the class shape by adding a method to one of the components. 
```

```typescript
@Component({
  selector: 'my-component',
  template: 'complex-template',
})
class SomeComponent {}

@Component({
  selector: 'my-component',
  template: 'empty-template',
})
class SomeMockedComponent {

>     public avoidCollision() {}

Reviewer 2: Can we do `private` here?
```suggestion
    private avoidCollision() {}
```
